### PR TITLE
Handle properly unnamed graphql queries when a graphql operation handler exists

### LINKED ISSUE #918: `graphql.operation` cannot intercept the unnamed graphql
## Environment

| Name    | Version |
| ------- | ------- |
| msw     | 0.35.0   |
| browser | Chrome  v93.0.4577.82   |
| OS      |   MacOS v11.5.2   |

## Request handlers

```js
import { setupWorker, rest } from 'msw'

const worker = setupWorker(
  graphql.operation((res, req, ctx) => {
    // ...
  })
)

worker.start()
```

## Actual request

```js
import { request, gql } from "graphql-request";

const query = gql`
  {
    hero {
      name
      friends {
        name
      }
    }
  }
`;

request("/", query)
```

## Current behavior

The handler didn't work, and there is a warning in the console:
```
[MSW] Failed to intercept a GraphQL request at "POST /": unnamed GraphQL operations are not supported.

Consider naming this operation or using "graphql.operation" request handler to intercept GraphQL requests regardless of their operation name/type. Read more: https://mswjs.io/docs/api/graphql/operation    
```
Note that using `rest.post` is workable:
```js
rest.post(/\//, (res, req, ctx) => {
   // ...
})
```

## Expected behavior

The handlers should works.

## Screenshots

Here is a reproducible [codesandbox](https://codesandbox.io/s/runtime-leftpad-okgvp?file=/src/App.js).

Finally, thanks for the fantastic library 🙌.  MSW helps my daily work a lot! Please let me know if there is anything I can help with this issue.

### PR DESCRIPTION
**Pull request by neilime**
- Fixes #918 

When registering a graphQL operation handler (https://mswjs.io/docs/api/graphql/operation), then send an unnamed graphQL query, we get this warning : 

```
Failed to intercept a GraphQL request at "${request.method} ${publicUrl}": unnamed GraphQL operations are not supported.
      
Consider naming this operation or using "graphql.operation" request handler to intercept GraphQL requests regardless of their operation name/type. Read more: https://mswjs.io/docs/api/graphql/operation
```

This behavior looks wrong.

This MR allows to send unnamed graphql queries when an "operation" handler exists

### CODE REVIEW

**Review by kettanaito — COMMENTED** [review]
Hey, @neilime. Thank you for reporting and fixing this!

From the first glance, it seems that `GraphQLHandler` isn't the correct surface to address this issue. That class shouldn't care what kind of higher-order handler (query/mutation/operation) is using it.

I'll give this a proper review once I have a chance and help you get this to a top-notch shape before merging. Thank you for contributing!

**Comment by neilime** [comment]
> Hey, @neilime. Thank you for reporting and fixing this!
> 
> From the first glance, it seems that `GraphQLHandler` isn't the correct surface to address this issue. That class shouldn't care what kind of higher-order handler (query/mutation/operation) is using it.
> 
> I'll give this a proper review once I have a chance and help you get this to a top-notch shape before merging. Thank you for contributing!

As the warning and the behaviour of "graphql.operation" leave in this GraphQLHandler class I was supposing that it was the best place to fix this issue.

BTW let me know how I can improve this MR, I will do it.

Actually I'm waiting for this fix, because it blocks my tests against an Apollo Federation Gateway that removes graphql Operation Name so I'm forced to use "graphql.operation"

File: src/handlers/GraphQLHandler.test.ts [hunk 1]
```diff
@@ -298,6 +306,15 @@ describe('predicate', () => {
     )
   })
 
+  test('respects graphql.operation endpoint', () => {
```
Comment (kettanaito) [inline src/handlers/GraphQLHandler.test.ts]: This is a great test but I think it should be added to the `test/graphql-api/operation.test.ts` test suite. That way the intention is closer to the feature (`graphql.operation`) that causes the issue.

Comment (kettanaito) [reply src/handlers/GraphQLHandler.test.ts]: This will also require restructuring the test. Look at the existing test cases in the suite to see how the GraphQL requests are performed (it'll be a browser context). Feel free to ask any questions here.

File: src/handlers/GraphQLHandler.ts [hunk 2]
```diff
@@ -211,4 +221,33 @@ Consider naming this operation or using "graphql.operation" request handler to i
...
+      : operationName === this.info.operationName
+  }
+
+  /**
+   * Checks if the handler is a "graphql.operation" handler.
+   * I.e accepts all type of operations
+   */
+  private isGraphqQLOperationHandler() {
+    return (
+      this.info.operationType === 'all' &&
```
Comment (kettanaito) [inline src/handlers/GraphQLHandler.ts]: I think it should be enough to rely on `this.info.operationType === 'all'` to reliably determine the `graphql.operation` handler, as it's its distinct value. We may drop the `isGraphQLOperationHandler` method altogether and adjust the error `if` statement to ensure we don't print the error when in the operation handler.

File: src/handlers/GraphQLHandler.ts [hunk 3]
```diff
@@ -211,4 +221,33 @@ Consider naming this operation or using "graphql.operation" request handler to i
     console.log('Response:', loggedResponse)
     console.groupEnd()
   }
+
+  /**
+   * Checks if the given operation name is matching with handler specifications
+   */
+  private hasMatchingOperationName(operationName: string | undefined) {
```
Comment (kettanaito) [inline src/handlers/GraphQLHandler.ts]: I'd refrain from any issue-unrelated refactoring to keep the pull request's scope minimal.

File: src/handlers/GraphQLHandler.ts [hunk 4]
```diff
@@ -162,30 +162,37 @@ export class GraphQLHandler<
       return false
     }
 
-    if (!parsedResult.operationName) {
```
Comment (kettanaito) [inline src/handlers/GraphQLHandler.ts:165]: I believe that the actual fix is simply checking that we're not in the `graphql.operation` handler:

```js
if (!parsedResult.operationName && this.info.operationType !== 'all') {
```

**Review by kettanaito — CHANGES_REQUESTED** [review]
Hey, @neilime. Thank you once more for this fix. I've left a few suggestions and would like to know your opinion on the proposed solution. Meanwhile, I'll add integration tests to `query` and `mutation` containing positive assertions that this warning is printed when indeed necessary.

File: src/handlers/GraphQLHandler.ts [hunk 5]
```diff
@@ -176,9 +176,10 @@ Consider naming this operation or using "graphql.operation" request handler to i
     const hasMatchingOperationType =
       this.info.operationType === 'all' ||
       parsedResult.operationType === this.info.operationType
+
     const hasMatchingOperationName =
       this.info.operationName instanceof RegExp
-        ? this.info.operationName.test(parsedResult.operationName)
+        ? this.info.operationName.test(parsedResult.operationName || '')
```
Comment (kettanaito) [inline src/handlers/GraphQLHandler.ts:182]: At this point in time `parsedResult.operationName` cannot be undefined, null, or an empty string. 
Was it some TypeScript error that made you add this fallback (`|| ''`)?

**Comment by kettanaito** [comment]
@neilime, may I please ask you to grant me permissions to push to the fork? I have a few minor touches to add before we merge this. Thanks.

**Comment by neilime** [comment]
> @neilime, may I please ask you to grant me permissions to push to the fork? I have a few minor touches to add before we merge this. Thanks.

@kettanaito Sure, done.

**Review by kettanaito — APPROVED** [review]
Thank you for fixing this, @neilime! 🎉

**Comment by kettanaito** [comment]
Welcome to contributors, @neilime!
